### PR TITLE
ICU-20854 Add support for more currency variants.

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20854
- [x] Updated PR title and link in previous line to include Issue number
- [ ] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by macchiati** [comment]
nice!

**Comment by sffc** [comment]
ICU-TC and Mark approved the API with some changes.  I need to address that feedback.

**Comment by sffc** [comment]
^ rebasing to master.

File: icu4j/main/classes/core/src/com/ibm/icu/text/CurrencyDisplayNames.java [hunk 1]
```diff
@@ -110,9 +110,10 @@ public static boolean hasData() {
     public abstract ULocale getULocale();
 
     /**
-     * Returns the symbol for the currency with the provided ISO code.  If
-     * there is no data for the ISO code, substitutes isoCode, or returns null
-     * if noSubstitute was set in the factory method.
+     * Returns the symbol for the currency with the provided ISO code.
+     *
+     * If there is no data for this symbol, substitutes isoCode,
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/text/CurrencyDisplayNames.java]: In javadoc (not C++ doxygen), start a new paragraph with `<p>`. No need to terminate it.

(I like to put it inline: ` * <p>If there is no...` but it's also ok to put it into the empty line.)

File: icu4j/main/classes/core/src/com/ibm/icu/text/CurrencyDisplayNames.java [hunk 2]
```diff
@@ -122,7 +123,12 @@ public static boolean hasData() {
 
     /**
      * Returns the narrow symbol for the currency with the provided ISO code.
-     * If there is no data for narrow symbol, substitutes the default symbol,
+     *
+     * The narrow currency symbol is similar to the regular currency symbol,
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/text/CurrencyDisplayNames.java]: ditto two paras here, and same for the new methods

File: icu4j/main/classes/core/src/com/ibm/icu/util/Currency.java [hunk 3]
```diff
@@ -90,14 +90,38 @@
 
     /**
      * Selector for getName() indicating the narrow currency symbol.
-     * The narrow currency symbol is similar to the regular currency
-     * symbol, but it always takes the shortest form: for example,
-     * "$" instead of "US$" for USD in en-CA.
+     *
+     * The narrow currency symbol is similar to the regular currency symbol,
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/util/Currency.java]: para...

File: icu4j/main/classes/core/src/com/ibm/icu/util/Currency.java [hunk 4]
```diff
@@ -90,14 +90,38 @@
...
+     * Selector for getName() indicating the formal currency symbol.
+     *
+     * The formal currency symbol is similar to the regular currency symbol,
+     * but it always takes the form used in formal settings such as banking;
+     * for example, "NT$" instead of "$" for TWD in zh-TW.
+     *
+     * @provisional This API might change or be removed in a future release.
+     * @draft ICU 67
+     */
+    public static final int FORMAL_SYMBOL_NAME = 4;
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/util/Currency.java]: I guess these are the "Shane to add integer constants to Currency.java." Please send an email addendum and update the proposal doc.

File: icu4j/main/classes/core/src/com/ibm/icu/util/Currency.java [hunk 5]
```diff
@@ -90,14 +90,38 @@
...
     public static final int NARROW_SYMBOL_NAME = 3;
 
+    /**
+     * Selector for getName() indicating the formal currency symbol.
+     *
+     * The formal currency symbol is similar to the regular currency symbol,
+     * but it always takes the form used in formal settings such as banking;
+     * for example, "NT$" instead of "$" for TWD in zh-TW.
+     *
+     * @provisional This API might change or be removed in a future release.
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/util/Currency.java]: FYI optional I guess -- we normally put draft first, then provisional. Tools probably don't care?

File: icu4j/main/classes/core/src/com/ibm/icu/impl/number/MutablePatternModifier.java [hunk 6]
```diff
@@ -400,10 +401,21 @@ public CharSequence getSymbol(int type) {
...
+                // Use currency.getName() here to make custom symbols work
+                return currency.getName(symbols.getULocale(), Currency.SYMBOL_NAME, null);
             } else {
-                int selector = unitWidth == UnitWidth.NARROW ? Currency.NARROW_SYMBOL_NAME
-                        : Currency.SYMBOL_NAME;
-                return currency.getName(symbols.getULocale(), selector, null);
+                CurrencyDisplayNames names = CurrencyDisplayNames.getInstance(symbols.getULocale());
+                switch (unitWidth) {
+                    case NARROW:
+                        return names.getNarrowSymbol(currency.getCurrencyCode());
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/number/MutablePatternModifier.java]: Question: Why getName() above "to make custom symbols work" but getSpecificXyzSymbol() here?

**Comment by markusicu** [comment]
AppVeyor: Cygwin lives... but fails? @jefgen 
```
make[1]: Making `all' in `toolutil'
make[1]: Making `check' in `toolutil'
make[2]: Entering directory '/cygdrive/c/projects/icu/icu4c/source/tools/toolutil'
rm -f ../../lib/libicutu.dll.a && ln -s libicutu66.dll.a ../../lib/libicutu.dll.a
make[2]: Entering directory '/cygdrive/c/projects/icu/icu4c/source/tools/toolutil'
rm -f ../../lib/libicutu.dll.a && ln -s libicutu66.dll.a ../../lib/libicutu.dll.a
ln: failed to create symbolic link '../../lib/libicutu.dll.a': File exists
make[2]: Leaving directory '/cygdrive/c/projects/icu/icu4c/source/tools/toolutil'
make[2]: *** [../../config/mh-cygwin64:108: ../../lib/cygicutu.dll] Error 1
make[2]: Leaving directory '/cygdrive/c/projects/icu/icu4c/source/tools/toolutil'
make[1]: Making `all' in `ctestfw'
make[1]: *** [Makefile:47: check-recursive] Error 2
```

**Comment by jefgen** [comment]
> AppVeyor: Cygwin lives... but fails? @jefgen Jeff Genovy FTE

I need to make a small change to the makefile for Cygwin 3.x to fix the parallel builds
I have a PR here: PR #1066 which fixes things. :)

**Comment by jefgen** [comment]
> ignore the AppVeyor error?
Yes, please ignore for now. thx

**Pull request merged**